url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+        getCore().raiseException(num, dbgmsg);
+    }
+
+    /**
+     * アドレシングモード 1 - データ処理オペランドを取得します。
+     *
+     * アセンブラでは shifter_operand と表されます。
+     *
+     * @param inst ARM 命令
+     * @return シフタオペランド
+     */
+    public int getAddrMode1(InstructionARM inst) {
+        boolean i = inst.getIBit();
+        boolean b7 = inst.getBit(7);
+        boolean b4 = inst.getBit(4);
+
+        if (i) {
+            //32bits イミディエート
+            return getAddrMode1Imm(inst);
+        } else if (!b4) {
+            //イミディエートシフト
+            return getAddrMode1ImmShift(inst);
+        } else if (b4 && !b7) {
+            //レジスタシフト
+            return getAddrMode1RegShift(inst);
+        } else {
+            throw new IllegalArgumentException("Unknown addr mode1 " +
+                    String.format("0x%08x, I:%b, b7:%b, b4:%b.",
+                            inst.getInst(), i, b7, b4));
+        }
+    }
+
+    /**
+     * アドレシングモード 1 - データ処理オペランド、
+     * 32ビットイミディエートを取得します。
+     *
+     * 条件:
+     * I ビット（ビット[25]）: 1
+     *
+     * rotate_imm: ビット[11:8]
+     * immed_8: ビット[7:0]
+     * とすると、イミディエート imm32 は下記のように求められます。
+     *
+     * imm32 = rotateRight(immed_8, rotate_imm * 2)
+     *
+     * @param inst ARM 命令
+     * @return イミディエート
+     */
+    public int getAddrMode1Imm(InstructionARM inst) {
+        int rotR = inst.getField(8, 4);
+        int imm8 = inst.getField(0, 8);
+
+        return Integer.rotateRight(imm8, rotR * 2);
+    }
+
+    /**
+     * アドレシングモード 1 - データ処理オペランド、
+     * イミディエートシフトを取得します。
+     *
+     * 条件:
+     * I ビット（ビット[25]）: 0
+     * ビット[4]: 0
+     *
+     * 該当するオペランド:
+     * 数値はビット[6:4] の値を示す。
+     * 0b000: データ処理オペランド - レジスタ
+     * 0b000: データ処理オペランド - イミディエート論理左シフト
+     * 0b010: データ処理オペランド - イミディエート論理右シフト
+     * 0b100: データ処理オペランド - イミディエート算術右シフト
+     * 0b110: データ処理オペランド - イミディエート右ローテート
+     * 0b110: データ処理オペランド - 拡張付き右ローテート
+     *
+     * @param inst ARM 命令
+     * @return イミディエートシフトオペランド
+     */
+    public int getAddrMode1ImmShift(InstructionARM inst) {
+        int shift_imm = inst.getField(7, 5);
+        int shift = inst.getField(5, 2);
+        int rm = inst.getRmField();
+
+        switch (shift) {
+        case 0:
+            if (shift_imm == 0) {
+                //レジスタ
+                return getReg(rm);
+            } else {
+                //イミディエート論理左シフト
+                return getReg(rm) << shift_imm;
+            }
+        case 1:
+            //イミディエート論理右シフト
+            if (shift_imm == 0) {
+                return 0;
+            } else {
+                return getReg(rm) >>> shift_imm;
+            }
+        case 2:
+            //イミディエート算術右シフト
+            if (shift_imm == 0) {
+                if (BitOp.getBit32(getReg(rm), 31)) {
+                    return 0xffffffff;
+                } else {
+                    return 0;
+                }
+            } else {
+                return getReg(rm) >> shift_imm;
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 199:207, 282:290.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.